consensus/subscribers.go

```diff
@@ -0,0 +1,86 @@
+package consensus
+
+// A Subscriber is an object that receives updates to the consensus set every
+// time there is a change in consensus.
+type Subscriber interface {
+	// ReceiveConsensusUpdate notifies subscribers that there has been a change in the
+	// consensus set, and provides an enumeration of the changes.
+	ReceiveConsensusUpdate(revertedBlocks []Block, appliedBlocks []Block)
+}
+
+// threadedSendUpdates sends updates to a specific subscriber as they become
+// available. One thread is needed per subscriber. A separate function was
+// needed due to race conditions; subscribers must receive updates in the
+// correct order. Furthermore, a deadlocked subscriber should not interfere
+// with consensus; updates cannot make blocking calls from any thread that is
+// holding a lock on consensus. The result is a construction where all updates
+// are added to a list of updates in the consensus set while the consensus set
+// is locked. Then, a separate thread for each subscriber will be notified (via
+// the update chan) that there are new updates. The thread will lock the
+// consensus set for long enough to get the updates, and then will unlock the
+// consensus set while it makes a blocking call to the subscriber. If the
+// subscriber deadlocks or has problems, the thread will stall indefinitely,
+// but the rest of consensus will not be disrupted.
+func (s *State) threadedSendUpdates(update chan struct{}, subscriber Subscriber) {
```

Comment:
hmm. Didn't we used to use a channel to serialize `AcceptBlock`? i.e. to enforce that only one thread was calling `AcceptBlock`? Maybe returning to that model would simply this.